java-checks/src/main/java/org/sonar/java/checks/serialization/SerialVersionUidInRecordCheck.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.serialization;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S6219")
+public class SerialVersionUidInRecordCheck extends IssuableSubscriptionVisitor {
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.RECORD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree targetRecord = (ClassTree) tree;
+    if (!targetRecord.symbol().type().isSubtypeOf("java.io.Serializable")) {
+      return;
+    }
+    for (Tree member : targetRecord.members()) {
+      if (!member.is(Tree.Kind.VARIABLE)) {
+        continue;
+      }
+      VariableTree variable = (VariableTree) member;
+      if (isSerialVersionUIDField(variable) && setsTheValueToZero(variable)) {
+        reportIssue(variable, "Remove this redundant \"serialVersionUID\" field");
+        return;
+      }
+    }
+  }
+
+  private static boolean isSerialVersionUIDField(VariableTree variable) {
+    Symbol symbol = variable.symbol();
+    return symbol.isFinal() &&
+      symbol.type().is("long") &&
+      symbol.name().equals("serialVersionUID");
+  }
+
+  private static boolean setsTheValueToZero(VariableTree variable) {
+    ExpressionTree initializer = variable.initializer();
+    if (initializer == null || !initializer.is(Tree.Kind.LONG_LITERAL)) {
```

Comment:
I would remove the check for `LONG_LITERAL` here. If the initializer is some other constant expression, like the name of another `static final long` in the same file, there's no reason we shouldn't still support it (unless I'm missing something). And if it's not, `asConstant` is going to return false anyway.

Side benefit: after removing this check, the check for `aLong.isPresent` can actually be false, so you'll increase test coverage.

Reply:
The issue is that I am not sure if this does not encourage [setting the value blindly](https://rules.sonarsource.com/java/tag/serialization/RSPEC-4926).

Basically, we can recover the value from:
1. A static variable
2. A static method

I can extend the current version to cover the first one but I am not sure I want to do the second one. WDYT?

Reply:
I don't think I understand your point about setting the value blindly. If we raise an issue in more cases, then, if anything, that'd discourage setting the value, no?

And regarding static methods: I'm just talking about removing the `.is` check and supporting whatever constant constructs `asConstant` already supports natively. Not supporting static methods is fine.